### ISSUE TITLE
Memory leak in tokio_util::sync::CancellationToken

### ISSUE THREAD

**Issue by Finomnis** [tokio-rs/tokio#4639]
**Version**
- tokio v1.17.0
- tokio-macros v1.7.0
- tokio-util v0.7.1

**Platform**

```
> uname -a
Linux WSL 5.10.102.1-microsoft-standard-WSL2 #1 SMP Wed Mar 2 00:30:59 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
> cat /etc/lsb-release`
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.4 LTS"
```

**Description**

I stumbled across a `Direct leak` in Rust's ASAN in one of my projects.

I reduced the code down to a minimal failing example:
```rust
use tokio_util::sync::CancellationToken;

fn main() {
    let parent = CancellationToken::new();
    let child = parent.child_token().child_token();
    parent.cancel();
    println!("{:?}", child);
}
```

**Error**

This is the full output of the previous command:
```
> cargo +nightly run --target x86_64-unknown-linux-gnu
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
     Running `target/x86_64-unknown-linux-gnu/debug/leak`
CancellationToken { is_cancelled: true }

=================================================================
==439==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 72 byte(s) in 1 object(s) allocated from:
    #0 0x5650b8b437fe in malloc /rustc/llvm/src/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:69:3
    #1 0x5650b8b7815b in alloc::alloc::alloc::h890a2995cad80481 /rustc/de1026a67b0a3f5d6b61a1f77093af97d4799376/library/alloc/src/alloc.rs:89:14
    #2 0x5650b8b78407 in alloc::alloc::Global::alloc_impl::hbff87b836a27dc2b /rustc/de1026a67b0a3f5d6b61a1f77093af97d4799376/library/alloc/src/alloc.rs:171:73
    #3 0x5650b8b790a0 in _$LT$alloc..alloc..Global$u20$as$u20$core..alloc..Allocator$GT$::allocate::h2b01323beaa914e3 /rustc/de1026a67b0a3f5d6b61a1f77093af97d4799376/library/alloc/src/alloc.rs:231:9
    #4 0x5650b8b77e66 in alloc::alloc::exchange_malloc::hcb6452b38bd9a654 /rustc/de1026a67b0a3f5d6b61a1f77093af97d4799376/library/alloc/src/alloc.rs:320:11
    #5 0x5650b8b6d025 in alloc::boxed::Box$LT$T$GT$::new::hb8724b9c226302f2 /rustc/de1026a67b0a3f5d6b61a1f77093af97d4799376/library/alloc/src/boxed.rs:200:9
    #6 0x5650b8b6d025 in tokio_util::sync::cancellation_token::CancellationToken::new::h89fe91bd7d43f37f /home/martin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-util-0.7.1/src/sync/cancellation_token.rs:144:21
    #7 0x5650b8b6bc90 in leak::main::h21ee5f4933260cfc /tmp/leak/src/main.rs:4:18
    #8 0x5650b8b6b10a in core::ops::function::FnOnce::call_once::ha737680e23d2f561 /rustc/de1026a67b0a3f5d6b61a1f77093af97d4799376/library/core/src/ops/function.rs:248:5
    #9 0x5650b8b6c403 in std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::h1ec3aa38eb5ffbf5 /rustc/de1026a67b0a3f5d6b61a1f77093af97d4799376/library/std/src/rt.rs:145:18
    #10 0x5650b8b6c24b in main (/tmp/leak/target/x86_64-unknown-linux-gnu/debug/leak+0x12224b) (BuildId: 311f261bad109925bc1d64fd07493de3d8894b4a)

SUMMARY: AddressSanitizer: 72 byte(s) leaked in 1 allocation(s).
```

**Reproduction**

At the time of writing, the address sanitizer is only available on Linux. (unsure about MacOS, but definitely not on Windows)

To reproduce the error, paste the code into the `main.rs` of a newly created Rust binary project.

- add the following to the `Cargo.toml`:
```toml
tokio = { version = "1.17.0", features = ["full"] }
tokio-util = { version = "0.7.1", default-features = false }
```

- install a nightly compiler toolchain
- enable address sanitizer by running:
```
export RUSTFLAGS="-Z sanitizer=address"
```

- compile and run
```
cargo +nightly run --target x86_64-unknown-linux-gnu
```

If you don't get the address resolution, you might have to install `llvm-symbolizer` (part of the default `llvm` packet in Ubuntu 20.04).

**Comment by Finomnis** [comment]
@Darksonn label should be A-tokio-utils instead of macros. I don't think I can modify it myself.

**Comment by Darksonn** [comment]
Ah, thanks. You're right.

**Comment by Darksonn** [comment]
Closed by #4652.
